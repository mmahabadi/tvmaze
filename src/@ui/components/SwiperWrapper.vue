<script setup lang="ts">
import { Swiper } from 'swiper/vue'
import { Autoplay, Pagination, Navigation } from 'swiper/modules'
import { type SwiperOptions } from '@/@ui/types'
import { defineProps, computed } from 'vue'

import 'swiper/css'
import 'swiper/css/autoplay'
import 'swiper/css/navigation'
import 'swiper/css/pagination'
import 'swiper/css/scrollbar'

const props = defineProps<Partial<SwiperOptions>>()

const swiperStyle = {
  '--swiper-navigation-color': 'rgb(20 184 166 / var(--tw-text-opacity))',
  '--swiper-pagination-color': 'rgb(20 184 166 / var(--tw-text-opacity))'
}
const swiperOptions = computed(() => ({
  ...props
}))

const convertToInt = (prop: number | undefined) => {
  return +(prop ?? 0)
}

const breakpoints = computed(() => ({
  0: {
    slidesPerView: Math.min(1, convertToInt(props.slidesPerView)),
    spaceBetween: Math.min(10, convertToInt(props.spaceBetween))
  },
  768: {
    slidesPerView: Math.min(2, convertToInt(props.slidesPerView)),
    spaceBetween: Math.min(10, convertToInt(props.spaceBetween))
  },
  1024: {
    slidesPerView: Math.min(3, convertToInt(props.slidesPerView)),
    spaceBetween: Math.min(20, convertToInt(props.spaceBetween))
  },
  1280: {
    slidesPerView: Math.min(5, convertToInt(props.slidesPerView)),
    spaceBetween: Math.min(30, convertToInt(props.spaceBetween))
  },
  1536: {
    slidesPerView: convertToInt(props.slidesPerView),
    spaceBetween: convertToInt(props.spaceBetween)
  }
}))
</script>
<template>
  <swiper
    :style="swiperStyle"
    :modules="[Autoplay, Pagination, Navigation]"
    v-bind="swiperOptions"
    :breakpoints="breakpoints"
  >
    <slot></slot>
  </swiper>
</template>
