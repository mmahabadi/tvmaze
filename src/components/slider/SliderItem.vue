<script setup lang="ts">
import { computed } from 'vue'
import DateIcon from '@/@ui/components/icons/CalendarIcon.vue'
import ClockIcon from '@/@ui/components/icons/ClockIcon.vue'
import IconPlay from '@/components/icons/IconPlay.vue'

const { name, image, endDate, rating, genre } = defineProps<{
  name: string
  image: string
  endDate: string
  rating: number
  genre: string
}>()

const year = computed(() => new Date(endDate).getFullYear())
</script>
<template>
  <img :src="image" alt="poster" class="w-full h-full object-cover object-top" />
  <div
    class="bg-black/40 absolute top-0 bottom-0 right-0 left-0 flex flex-col justify-center items-center"
  >
    <div
      class="details transition-opacity duration-200 ease-in opacity-0 flex flex-col justify-center md:gap-5 gap-4 items-center"
    >
      <h1 class="xl:text-4xl truncate capitalize font-sans sm:text-2xl text-4xl font-bold">
        {{ name }}
      </h1>
      <div class="flex gap-5 items-center">
        <div class="flex items-center gap-2">
          <span class="text-sm font-medium">{{ genre }}</span>
        </div>
        <div class="flex items-center gap-2">
          <DateIcon class="text-teal-500" /><span class="text-sm font-medium">{{ year }}</span>
        </div>
        <div class="flex items-center gap-2">
          <ClockIcon class="text-teal-500" /><span class="text-sm font-medium">{{ rating }}hr</span>
        </div>
      </div>
      <div class="flex gap-5 items-center">
        <a
          class="watch-button bg-teal-500 hover:bg-teal-700 text-white font-bold py-2 px-4 rounded inline-flex items-center"
          :href="name"
        >
          <IconPlay class="text-white mr-2" />
          <IconPlay class="text-white mr-2 opacity-0 absolute animate-ping" />
          Watch
        </a>
      </div>
    </div>
  </div>
</template>
<style scoped>
.swiper-slide-next .details,
.watch-button:hover .animate-ping {
  opacity: 100;
}
</style>
