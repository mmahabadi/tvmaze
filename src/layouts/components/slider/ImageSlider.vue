<template>
  <!-- <div
    class="swiper swiper-initialized swiper-vertical swiper-pointer-events w-full xl:h-96 bg-dry lg:h-64 h-48 swiper-backface-hidden"
  ></div> -->
  <swiper
    :modules="modules"
    :slides-per-view="1"
    :space-between="50"
    Autoplay
    @swiper="onSwiper"
    @slideChange="onSlideChange"
  >
    <swiper-slide>Slide 1</swiper-slide>
    <swiper-slide>Slide 2</swiper-slide>
    <swiper-slide>Slide 3</swiper-slide>
    <swiper-slide>Slide 4</swiper-slide>
    <swiper-slide>Slide 5</swiper-slide>
    <swiper-slide>Slide 6</swiper-slide>
  </swiper>
</template>
<script>
// Import Swiper Vue.js components
import { Swiper, SwiperSlide } from 'swiper/vue'

import 'swiper/css'
import 'swiper/css/autoplay'

export default {
  components: {
    Swiper,
    SwiperSlide
  },
  setup() {
    const onSwiper = (swiper) => {
      console.log(swiper)
    }
    const onSlideChange = () => {
      console.log('slide change')
    }
    return {
      onSwiper,
      onSlideChange
    }
  }
}
</script>
