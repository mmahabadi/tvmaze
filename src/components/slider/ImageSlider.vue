<script setup lang="ts">
import { SwiperSlide } from 'swiper/vue'
import { storeToRefs } from 'pinia'
import { useMoviesStore } from '@/stores/movies'
import SliderItem from './SliderItem.vue'
import SwiperWrapper from '@/@ui/components/SwiperWrapper.vue'

import 'swiper/css'
import 'swiper/css/autoplay'

const store = useMoviesStore()
const { recentMovies } = storeToRefs(store)

const options = {
  slidesPerView: 3,
  spaceBetween: 0,
  speed: 600,
  lazy: true,
  autoplay: {
    delay: 5000,
    disableOnInteraction: true
  }
} as SwiperOptions
</script>
<template>
  <SwiperWrapper v-bind="options" class="w-full -mt-24 h-[44rem]">
    <swiper-slide v-for="item in recentMovies" :key="item.id">
      <SliderItem
        :name="item.name"
        :image="item.image?.original"
        :endDate="item.ended"
        :rating="item.rating?.average"
        :genre="item.genres?.[0]"
      />
    </swiper-slide>
  </SwiperWrapper>
</template>
