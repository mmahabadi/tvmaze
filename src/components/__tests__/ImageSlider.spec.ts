import { describe, it, expect, beforeEach } from 'vitest'
import { mount, VueWrapper } from '@vue/test-utils'
import ImageSlider from '../slider/ImageSlider.vue'
import SliderItem from '../slider/SliderItem.vue'
import SwiperWrapper from '@/@ui/components/SwiperWrapper.vue'
import { createPinia, setActivePinia } from 'pinia'

describe('ImageSlider', () => {
  let wrapper: VueWrapper<ImageSlider>

  beforeEach(() => {
    setActivePinia(createPinia())
    wrapper = mount(ImageSlider, {
      global: {
        components: {
          SliderItem,
          SwiperWrapper
        }
      }
    })
  })

  it('renders correctly', () => {
    console.log(wrapper.html())
    expect(wrapper.html()).toMatchSnapshot()
  })
})
