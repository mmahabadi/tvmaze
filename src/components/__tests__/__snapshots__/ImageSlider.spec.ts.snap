// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ImageSlider > renders correctly 1`] = `
"<div class="swiper swiper-initialized swiper-horizontal w-full -mt-24 h-[44rem]" style="--swiper-navigation-color: var(--primary-color); --swiper-pagination-color: var(--primary-color);" lazy="true" data-textid="slider">
  <div class="swiper-wrapper" style="transition-duration: 0ms; transition-delay: 0ms; transform: translate3d(NaNpx, 0px, 0px);"></div>
  <!---->
  <!---->
  <!---->
</div>"
`;
